File: tree_primer/tree_primer__center_of_tree.py
# 実行時間 20ms
n = int(input())
v = n
e = [True] * n
g = [[False] * n for _ in [0] * n]

for _ in [0] * (n - 1):
    a, b = list(map(int, input().split()))

    a -= 1
    b -= 1
    g[a][b] = True
    g[b][a] = True
while 2 < v:
    for i in range(n):
        cnt = 0

        for j in range(n):
            if g[i][j] == True:
                cnt += 1
        if cnt == 1:
            v -= 1
            e[i] = False
            g[i] = [False] * n
    for i in range(n):
        for j in range(n):
            if e[j] != True and g[i][j] == True:
                g[i][j] = False
for i in range(n):
    if e[i] == True:
        print(i + 1)

File: tree_primer/tree_primer__center_of_tree.cs
// 実行時間 20ms
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        int n = int.Parse( ReadLine() );
        int v = n;
        bool[] e = Repeat( true, n ).ToArray();
        bool[][] g = new bool[n][];
        var rng = Range( 0, n );

        foreach( int i in rng ) {
            g[i] = new bool[n];
        }
        foreach( int _ in Range( 0, n - 1 ) ) {
            int[] a = ReadLine().Trim().Split().Select( x => int.Parse( x ) - 1 ).ToArray();

            g[a[0]][a[1]] = true;
            g[a[1]][a[0]] = true;
        }
        while( 2 < v ) {
            foreach( int i in rng ) {
                int cnt = 0;

                foreach( int j in rng ) {
                    if( g[i][j] == true ) {
                        cnt++;
                    }
                }
                if( cnt == 1 ) {
                    v--;
                    e[i] = false;
                    g[i] = new bool[n];
                }
            }
            foreach( int i in rng ) {
                foreach( int j in rng ) {
                    if( e[j] != true && g[i][j] == true ) {
                        g[i][j] = false;
                    }
                }
            }
        }
        foreach( int i in rng.Where( i => e[i] == true ) ) {
            WriteLine( i + 1 );
        }
        return;
    }
}
